src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRefreshListener.cs

```diff
@@ -45,9 +47,27 @@ internal class SemanticTokensRefreshListener
 
         private void OnLspWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e) => _workQueue.AddWork();
 
-        // TO-DO: Replace hardcoded string with const once LSP side is merged.
-        public ValueTask SendSemanticTokensNotificationAsync(CancellationToken cancellationToken)
-            => _languageServerNotificationManager.SendNotificationAsync(Methods.WorkspaceSemanticTokensRefreshName, cancellationToken);
+        public async ValueTask SendSemanticTokensNotificationAsync(CancellationToken cancellationToken)
+        {
+            // Currently, we send a refresh notification to the client whenever there are any workspace changes.
+            // However, we also need to send notifications whenever the compilations finish processing since
+            // compilations impact semantic tokens. We put this logic here rather than in the semantic tokens
```

Comment:
I wonder - can this still go into the range handler itself?  Basically when we get a semantic tokens range request we trigger the compilation request (if not available) which triggers an event of some kind that feeds here?

That way we get de-bouncing as well on our side.